runtime/from-json-processor.h

```diff
@@ -0,0 +1,217 @@
+    : json_(json)
+    , flatten_class_(flatten_class)
+    , json_path_ (json_path) {}
+
+  template<class T>
+  void operator()(const char *key, T &value, bool required = false) noexcept {
+    if (!error_.empty()) {
+      return;
+    }
+    if (flatten_class_) {
+      do_set(value, json_);
+      return;
+    }
+    json_path_.enter(key);
+    const auto *json_value = json_.as_array().find_value(string{key});
+    if (required && !json_value) {
+      error_.append("absent required field ");
+      error_.append(json_path_.to_string());
+    }
+    if (json_value) {
+      do_set(value, *json_value);
+    }
+    json_path_.leave();
+  }
+
+  bool has_error() const noexcept { return !error_.empty(); }
+  const string &get_error() const noexcept { return error_; }
+
+  static const string &get_json_obj_magic_key() noexcept {
+    static string key{"__json_obj_magic"};
+    return key;
+  }
+
+private:
+  [[gnu::noinline]] void on_input_type_mismatch(const mixed &json) noexcept {
+     error_.assign("unexpected type ");
+     if (json.is_array()) {
+       error_.append(json.as_array().is_vector() ? "array" : "object");
+     } else {
+       error_.append(json.get_type_str());
+     }
+     error_.append(" for key ");
+     error_.append(json_path_.to_string());
+  }
+
+  void do_set(bool &value, const mixed &json) noexcept {
+    if (!json.is_bool()) {
+      on_input_type_mismatch(json);
+      return;
+    }
+    value = json.as_bool();
+  }
+
+  void do_set(std::int64_t &value, const mixed &json) noexcept {
+    if (!json.is_int()) {
+      on_input_type_mismatch(json);
+      return;
+    }
+    value = json.as_int();
+  }
+
+  void do_set(double &value, const mixed &json) noexcept {
+    if (!json.is_float() && !json.is_int()) {
+      on_input_type_mismatch(json);
+      return;
+    }
+    value = json.as_double();
+  }
+
+  void do_set(string &value, const mixed &json) noexcept {
+    if (!json.is_string()) {
+      on_input_type_mismatch(json);
+      return;
+    }
+    value = json.as_string();
+  }
+
+  void do_set(JsonRawString &value, const mixed &json) noexcept {
+    static_SB.clean();
+    if (!impl_::JsonEncoder{0, false, get_json_obj_magic_key()}.encode(json)) {
+      error_.append("failed to decode @kphp-json raw_string field ");
+      error_.append(json_path_.to_string());
+      return;
+    }
+    value.str = static_SB.str();
+  }
+
+  template<class T>
+  void do_set(Optional<T> &value, const mixed &json) noexcept {
+    if (json.is_null()) {
+      value = Optional<bool>{};
+      return;
+    }
+    do_set(value.ref(), json);
+  }
+
+  template<class I>
+  void do_set(class_instance<I> &klass, const mixed &json) noexcept;
+
+  // just don't fail compilation with empty untyped arrays
+  void do_set(array<Unknown> &/*array*/, const mixed &/*json*/) noexcept {}
+
+  template<class T>
+  void do_set_array(array<T> &array, const mixed &json) noexcept {
+    const auto &json_array = json.as_array();
```

Comment:
let's add an assertion that it's an array

Reply:
It's called from `if (json.is_array())`